### PR TITLE
fix isDisplayed overflow-x/y bug

### PR DESCRIPTION
**Pull request by louisremi**
## Proposed changes

Fix regression inside isDisplay introduced with version 8.2.22: Webdriverio now incorrectly reports elements with style `overflow-x: hidden; width: 0` as displayed.
This fixes https://github.com/webdriverio/webdriverio/issues/12781

## Types of changes

- [ ] Polish (an improvement for an existing feature)
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update (improvements to the project's docs)
- [ ] Specification changes (updates to WebDriver command specifications)
- [ ] Internal updates (everything related to internal scripts, governance documentation and CI files)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] ~~I have added the necessary documentation (if appropriate)~~
- [ ] ~~I have added proper type definitions for new commands (if appropriate)~~

## Backport Request

[//]: # (The current `main` branch is the development branch for WebdriverIO v9. If your change should be released to the current major version of WebdriverIO (v8), please raise another PR with the same changes against the `v8` branch.)

- [ ] This change is solely for `v9` and doesn't need to be back-ported
- [ ] Back-ported PR at `#XXXXX`

## Further comments

This PR is a Work In Progress:
- [x] Add more tests for `opacity: 0`, `visibility: hidden` and `display: none` to the test suite
- [ ] Decide whether to keep fixture inside test file as Lit component, or to add them to guinea-pig
- [ ] Backport to v8

### CODE REVIEW

**Review by christian-bromann — COMMENTED** [review]
Thanks for taking a stab at this. I have one question on the code changes. Regarding testing, you currently use the component test runner for this which is fine. An alternative is to make a PR to https://github.com/webdriverio/guinea-pig which is a simple website accessible at http://guinea-pig.webdriver.io/. This can be used to then to just keep this as e2e test, use the `url` command instead of `render` and move the test somewhere `e2e/wdio/headless/test.e2e.ts`.

File: e2e/browser-runner/__fixtures__/is-displayed.html [hunk 1]
```diff
@@ -0,0 +1,36 @@
+<head>
```
[the remark this answers is not in the document]
Comment (louisremi) [reply e2e/browser-runner/__fixtures__/is-displayed.html]: Alright, I removed that file

File: e2e/browser-runner/commands/element/isDisplayed.test.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+// import { screen } from '@testing-library/jest-dom'
+// import userEvent from '@testing-library/user-event'
+import { html, render } from 'lit'
+
+import * as matchers from '@testing-library/jest-dom/matchers'
+expect.extend(matchers)
+
+describe('isDisplayed', () => {
+    beforeEach(async () => {
+        // TODO: this fails with "Failed to load test page […]" error
```
[the remark this answers is not in the document]
Comment (louisremi) [reply e2e/browser-runner/commands/element/isDisplayed.test.ts]: This is no longer a problem, I got rid of the fixture file. I render a document with Lit for now, and I can submit a PR to guinea pig if you prefer

File: e2e/package.json [hunk 3]
```diff
@@ -5,6 +5,9 @@
   "private": true,
   "scripts": {
     "test:browser": "run-s test:browser:*",
+    "test:browser:isDisplayed": "run-s test:browser:isDisplayed:*",
+    "test:browser:isDisplayed:wdio": "cross-env WDIO_PRESET=isDisplayed node ../packages/wdio-cli/bin/wdio.js ./browser-runner/wdio.conf.js --spec isDisplayed.test.ts",
+    "test:browser:isDisplayed:verifyCoverage": "run-s verifyCoverage",
```
[the remark this answers is not in the document]
Comment (louisremi) [reply e2e/package.json]: I got rid of this line

File: packages/webdriverio/src/scripts/isElementDisplayed.ts [hunk 4]
```diff
@@ -266,11 +238,11 @@ export default function isElementDisplayed (element: Element): boolean {
         return false
     }
 
-    if (!elementSubtreeHasNonZeroDimensions(element)) {
+    if (!elementSubtreeHasNonZeroDimension({ element, dimension: 'x' })) {
         return false
     }
 
-    if (isElementSubtreeHiddenByOverflow(element) && !elementHasBoundingBox(element)) {
+    if (!elementSubtreeHasNonZeroDimension({ element, dimension: 'y' })) {
```
[the remark this answers is not in the document]
Comment (louisremi) [reply packages/webdriverio/src/scripts/isElementDisplayed.ts:245]: Yes, there are various tests in the test-suite I created that check the result of only using `overflow` instead of `overflow-x/y`

File: packages/webdriverio/src/scripts/isElementDisplayed.ts [hunk 5]
```diff
@@ -114,25 +114,38 @@ export default function isElementDisplayed (element: Element): boolean {
         return cascadedStylePropertyForElement(parentElement, property)
     }
 
-    function elementHasBoundingBox(element: Element): boolean {
+    function elementSubtreeHasNonZeroDimension({
```
Comment (erwinheitzman) [inline packages/webdriverio/src/scripts/isElementDisplayed.ts:117]: It's better to check for Zero dimensions than for non Zero dimensions for the function. Names like these are confusing especially when executed like so !nonZero

File: packages/webdriverio/src/scripts/isElementDisplayed.ts [hunk 6]
```diff
@@ -114,25 +114,38 @@ export default function isElementDisplayed (element: Element): boolean {
...
         }
 
         // Paths can have a zero width or height. Treat them as shown if the stroke width is positive.
-        const boundingBox = element.getBoundingClientRect()
-        if (element.tagName.toUpperCase() === 'PATH' && boundingBox.width + boundingBox.height > 0) {
-            const strokeWidth = cascadedStylePropertyForElement(element, 'stroke-width')
-            return !!strokeWidth && (parseInt(strokeWidth, 10) > 0)
+        if (
+            element.tagName.toUpperCase() === 'PATH' &&
+            (boundingBox.width > 0 || boundingBox.height > 0)
```
Comment (erwinheitzman) [inline packages/webdriverio/src/scripts/isElementDisplayed.ts:132]: Shouldn't this be && instead of ||?

File: packages/webdriverio/src/scripts/isElementDisplayed.ts [hunk 7]
```diff
@@ -144,58 +157,16 @@ export default function isElementDisplayed (element: Element): boolean {
...
-        if (!elementOverflowsContainer(element)) {
-            return false
-        }
-
-        if (!element.childNodes.length) {
-            return false
-        }
-
-        // This element's subtree is hidden by overflow if all child subtrees are as well.
-        return [].every.call(element.childNodes, function (childNode: Element) {
```
Comment (erwinheitzman) [inline packages/webdriverio/src/scripts/isElementDisplayed.ts:180]: We lose this functionality by removing this, the purpose of this is to check if any child elements of an element are not displayed. We need to keep this unless I missed something

**Review by erwinheitzman — COMMENTED** [review]
Although I like the changes to support the mentioned surface, I do not feel confident in merging this. Especially since changing this code potentially has A LOT of impact on our users, so getting it right is very important. I would like to run this on a big codebase first before merging this. Thank you very much for looking into this and contributing, great find overall 🙂

File: packages/webdriverio/src/scripts/isElementDisplayed.ts (hunk 6 above)
Comment (louisremi) [reply packages/webdriverio/src/scripts/isElementDisplayed.ts:132]: This replaces `boundingBox.width + boundingBox.height > 0` for clarity, so `||` is appropriate

File: packages/webdriverio/src/scripts/isElementDisplayed.ts (hunk 5 above)
Comment (louisremi) [reply packages/webdriverio/src/scripts/isElementDisplayed.ts:117]: The diff makes it seem that this is new code, while it isn't: this is an existing function that I modified to check dimensions one by one instead. 
The function is used recursively. I agree that at the top level it would be better to write `elementSubtreeHasZeroDimension()" instead of "!elementSubtreeHasNonZeroDimension()`. But in the recursive algorithm, the original name makes more sense. We could write a new `const elementSubtreeHasZeroDimension = (...args) => !elemelementSubtreeHasNonZeroDimension(...args)` helper, but I feel that would be a bit overkill

File: packages/webdriverio/src/scripts/isElementDisplayed.ts (hunk 7 above)
Comment (louisremi) [reply packages/webdriverio/src/scripts/isElementDisplayed.ts:180]: After a careful analysis, it seemed that this test `isElementSubtreeHiddenByOverflow` was redundant with the test `elementSubtreeHasNonZeroDimensions`. 
I'm not sure why those two different tests existed in the first place, but `isElementSubtreeHiddenByOverflow` seemed like an inferior version (not testing for path with > 0 stroke, weird recursive implementation). So I got rid of it.
But I agree this change should be carefully tested on a large code-base before being validated.

**Comment by erwinheitzman** [comment]
@louisremi I would really appreciate it if you could do that but I do not expect you to go that far (perhaps nice to make sure the fix works on your end though).

My plan is to run this against all the tests of the client I currently work for. It's scale is huge as well and it should give us a good indication of any impact on existing testcases. I do need to find time for it though

**Comment by christian-bromann** [comment]
@louisremi @erwinheitzman any updates on this one?

**Comment by christian-bromann** [comment]
Closing due to inactivity.

**Comment by christian-bromann** [comment]
@erwinheitzman did you had a chance to test this in a larger code base?

**Comment by christian-bromann** [comment]
Closing in favor of #14075
